mavros_extras/src/plugins/esc_status.cpp

```diff
@@ -0,0 +1,148 @@
+/**
+ * @brief ESC status plugin
+ * @file esc_status.cpp
+ * @author Ricardo Marques <marques.ricardo17@gmail.com>
+ *
+ * @addtogroup plugin
+ * @{
+ */
+/*
+ * Copyright 2020 Ricardo Marques <marques.ricardo17@gmail.com>.
+ *
+ * This file is part of the mavros package and subject to the license terms
+ * in the top-level LICENSE file of the mavros repository.
+ * https://github.com/mavlink/mavros/tree/master/LICENSE.md
+ */
+
+#include <mavros/mavros_plugin.h>
+#include <mavros_msgs/ESCInfo.h>
+#include <mavros_msgs/ESCStatus.h>
+
+namespace mavros
+{
+namespace extra_plugins
+{
+/**
+ * @brief ESC status plugin
+ */
+class ESCStatusPlugin : public plugin::PluginBase
+{
+public:
+	ESCStatusPlugin() : PluginBase(),
+		_max_esc_info_index(0),
+		_max_esc_status_index(0),
+		nh("~")
+	{}
+
+	void initialize(UAS &uas_)
+	{
+		PluginBase::initialize(uas_);
+
+		esc_info_pub = nh.advertise<mavros_msgs::ESCInfo>("esc_info", 10);
+		esc_status_pub = nh.advertise<mavros_msgs::ESCStatus>("esc_status", 10);
+
+		enable_connection_cb();
+	}
+
+	Subscriptions get_subscriptions()
+	{
+		return {
+			       make_handler(&ESCStatusPlugin::handle_esc_info),
+			       make_handler(&ESCStatusPlugin::handle_esc_status),
+		};
+	}
+
+private:
+	using lock_guard = std::lock_guard<std::mutex>;
+	std::mutex mutex;
+
+	ros::NodeHandle nh;
+
+	ros::Publisher esc_info_pub;
+	ros::Publisher esc_status_pub;
+	mavros_msgs::ESCInfo _esc_info;
+	mavros_msgs::ESCStatus _esc_status;
+	uint8_t _max_esc_info_index;
+	uint8_t _max_esc_status_index;
+	const uint8_t batch_size = 4;
+
+	void handle_esc_info(const mavlink::mavlink_message_t *msg, mavlink::common::msg::ESC_INFO &esc_info)
+	{
+		lock_guard lock(mutex);
+
+		_esc_info.header.stamp = m_uas->synchronise_stamp(esc_info.time_usec);
+
+		uint8_t esc_index = esc_info.index;
+
+		_esc_info.counter = esc_info.counter;
+		_esc_info.count = esc_info.count;
+		_esc_info.connection_type = esc_info.connection_type;
+		_esc_info.info = esc_info.info;
+
+		size_t required_size = _esc_info.count;
+		if (_esc_info.esc_info.size() < required_size)
+		{
+			_esc_info.esc_info.resize(_esc_info.count);
+		}
+
+		for (int i = 0; i < batch_size; i++)
+		{
+			_esc_info.esc_info[esc_index + i].header =  _esc_info.header;
+			_esc_info.esc_info[esc_index + i].failure_flags = esc_info.failure_flags[i];
+			_esc_info.esc_info[esc_index + i].error_count = esc_info.error_count[i];
+			_esc_info.esc_info[esc_index + i].temperature = esc_info.temperature[i];
+		}
+
+		_max_esc_info_index = std::max(_max_esc_info_index, esc_info.index);
+
+		if (_max_esc_info_index == esc_info.index)
+		{
+			esc_info_pub.publish(_esc_info);
+		}
+	}
+
+	void handle_esc_status(const mavlink::mavlink_message_t *msg, mavlink::common::msg::ESC_STATUS &esc_status)
+	{
+		lock_guard lock(mutex);
+
+		uint8_t esc_index = esc_status.index;
+
+		size_t required_size = batch_size * (esc_index ) + batch_size;	// TODO FIX. We should only add as many fields as they exist.
```

Comment:
@vooon How should we proceed here. I can see several options but not really the "optimal" one.

1. Leave as is. It means that if we are receiving messages from a Hexacopter we will be publishing vector size 8 when size 6 would suffice. Pros: Simple and foolproof.

2. Incorporate the data from the `esc_info` message and so: `_esc_info.esc_info.resize(_esc_info.count);` . Cons: It requires a second message and if a user is only sending the `ESC_STATUS` mavlink message this won't really work.

3. Introduce a count field in the `ESC_STATUS` mavlink index. Cons: Mavlink message gets bigger but not that much tbf. It just requires a change from the start but it can be done.

4. Analyze the data fields in the `ESC_STATUS` message. This should be defined from the MAVLink side but if that field is empty it should have the max value. I guess in this case to be ultra safe we should check this for both `rpm` `voltage` and `current` . This will prevent messages being dropped just because a certain field is missing.

5. ??

Let me know what you think.

Reply:
Umm, i thought that `esc_info.count` represent count of that particular message (wrong assumption based on older massages).

So yes, for esc_info vector we just `.resize(count)`. Then we should loop `i=0; i < min(count - 4*index, 4); i++`.

Reply:
And we should use esc_info.count to process esc_status too. As both messages should have equal amount of items.
Status parser should check that message can fit into the vector, otherwise drop message with warning.

Reply:
Sorry @vooon this became a bit confusing as with the other message there seems to be a slight confusion here with count, index etc

Should I just store in a variable the `_max_esc_count` define it in the `handle_esc_info` and then also use it in the `handle_esc_status` ? I.e. suggestion 2 in my comment? This assumes that both messages are always being published however, hence my question to you.

Reply:
Yes, just store total number of escs. I assume that usage of only ESC_STATUS is agains spec, so we shouldn't think host to support that.

Reply:
Sounds good. I'll do that.

However if you feel like suggestion number 3 is a bit more robust let me know and I can work towards that after this PR is merged. The mavlink messages are WIP so this is one of those cases where we can improve it after the fact.